src/types/quota.rs

```diff
@@ -0,0 +1,246 @@
+use crate::error::{Error, ParseError};
+use crate::parse::{parse_many_into2, parse_until_done, MapOrNot, MapOrNot2};
+use crate::types::UnsolicitedResponse;
+use imap_proto::Response;
+use ouroboros::self_referencing;
+use std::borrow::Cow;
+use std::fmt::{Debug, Display, Formatter};
+use std::sync::mpsc;
+
+/// From [SETQUOTA Resource limit](https://datatracker.ietf.org/doc/html/rfc2087#section-4.1)
+///
+/// Used by [`Session::set_quota`]
+#[derive(Debug, Eq, PartialEq)]
+#[non_exhaustive]
+pub struct QuotaResourceLimit<'a> {
+    /// The resource type
+    pub name: QuotaResourceName<'a>,
+    /// The amount for that resource
+    pub amount: u64,
+}
+
+impl<'a> QuotaResourceLimit<'a> {
+    /// Creates a new [`QuotaResourceLimit`]
+    pub fn new(name: QuotaResourceName<'a>, amount: u64) -> Self {
+        Self { name, amount }
+    }
+}
+
+impl Display for QuotaResourceLimit<'_> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{} {}", self.name, self.amount)
+    }
+}
+
+/// From [Resources](https://datatracker.ietf.org/doc/html/rfc2087#section-3)
+///
+/// Used by [`QuotaLimit`], and [`QuotaResource`]
+#[derive(Debug, Eq, PartialEq, Clone)]
+#[non_exhaustive]
+pub enum QuotaResourceName<'a> {
+    /// Sum of messages' RFC822.SIZE, in units of 1024 octets
+    Storage,
+    /// Number of messages
+    Message,
+    /// Any other string (for future RFCs)
+    Atom(Cow<'a, str>),
+}
+
+impl Display for QuotaResourceName<'_> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::Storage => write!(f, "STORAGE"),
+            Self::Message => write!(f, "MESSAGE"),
+            Self::Atom(s) => write!(f, "{}", s),
+        }
+    }
+}
+
+impl<'a> QuotaResourceName<'a> {
+    /// Get an owned version of the [`QuotaResourceName`].
+    pub fn into_owned(self) -> QuotaResourceName<'static> {
+        match self {
+            QuotaResourceName::Storage => QuotaResourceName::Storage,
+            QuotaResourceName::Message => QuotaResourceName::Message,
+            QuotaResourceName::Atom(n) => QuotaResourceName::Atom(Cow::Owned(n.into_owned())),
+        }
+    }
+}
+
+/// From [QUOTA Response](https://datatracker.ietf.org/doc/html/rfc2087#section-5.1)
+///
+/// This is a wrapper around a single single [`Quota`].
+///
+/// Used by [`Session::get_quota`] and [`Session::set_quota`]
+#[self_referencing]
+pub struct QuotaResponse {
+    data: Vec<u8>,
+    #[borrows(data)]
+    #[covariant]
+    pub(crate) quota: Option<Quota<'this>>,
+}
+
+impl QuotaResponse {
+    /// Parse the [`Quota`] response from a response buffer.
+    pub fn parse(
```

Comment:
Does the parser need to be `pub`? Same for the other `fn parse` in here.

Reply:
I only left it "pub" as all existing parsers were put. 

Though having it "pub" was actually helpful in writing tests for my app that uses this crate as I could easily write mocks to build the responses of the imap methods. (if they were not pub there would be no way for me to do that).

I have a mock_double for the imap connection so I can test my code and stub the responses (e.g. having my tests not rely on having an actual IMAP server there).

IMHO we can address this in a separate PR e.g. building a "testing module" or something to allow easy building of those ouroboros response objects.

Reply:
A separate PR sounds like a good idea. I do think that in general the parsing methods should probably not be public, or at the very least be in their own module.